pallets/migrations/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2019-2020 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Migration Pallet
+
+#![allow(non_camel_case_types)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_support::{pallet, weights::Weight};
+
+pub use pallet::*;
+
+#[cfg(test)]
+#[macro_use]
+extern crate environmental;
+
+/// A Migration that must happen on-chain upon a runtime-upgrade
+pub trait Migration {
+	/// A human-readable name for this migration. Also used as storage key.
+	fn friendly_name(&self) -> &str;
+
+	/// Perform the required migration and return the weight consumed.
+	///
+	/// Currently there is no way to migrate across blocks, so this method must (1) perform its full
+	/// migration and (2) not produce a block that has gone over-weight. Not meeting these strict
+	/// constraints will lead to a bricked chain upon a runtime upgrade because the parachain will
+	/// not be able to produce a block that the relay chain will accept.
+	fn migrate(&self, available_weight: Weight) -> Weight;
+}
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_std::prelude::*;
+
+	/// Pallet for migrations
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The list of migrations that will be performed
+		type MigrationsList: Get<Vec<Box<dyn Migration>>>;
```

Comment:
I agree with an earlier suggestion to change this trait bound to `Migration` and allow it to accept multiple migrations by implementing `Migration` for **tuple** instead of passing them in as a vec.

Reply:
Agree. We can do it in a followup though.